services/circleci/circleci.service.js

```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+const Joi = require('joi')
+const BaseJsonService = require('../base-json')
+
+const circleSchema = Joi.array()
+  .items(Joi.object({ status: Joi.string() }))
+  .min(1)
+  .required()
+
+module.exports = class CircleCi extends BaseJsonService {
+  async fetch({ token, type, userRepo, branch }) {
+    let url = `https://circleci.com/api/v1.1/project/${type}/${userRepo}`
+    if (branch != null) {
+      url += `/tree/${branch}`
+    }
+    const query = { filter: 'completed', limit: 1 }
+    if (token) {
+      query['circle-token'] = token
+    }
+    return this._requestJson({
+      url,
+      schema: circleSchema,
+      options: { qs: query },
+      errorMessages: { 404: 'project not found' },
+    })
+  }
+
+  static render({ status, color }) {
+    return { message: status, color: color }
+  }
+
+  transform(data) {
+    let passCount = 0
+    let circleStatus, shieldsStatus, color
+
+    for (let i = 0; i < data.length; i++) {
+      circleStatus = data[i].status
+      if (['success', 'fixed'].includes(circleStatus)) {
+        passCount++
+      } else if (circleStatus === 'failed') {
+        shieldsStatus = 'failed'
+        color = 'red'
+        return { status: shieldsStatus, color: color }
+      } else if (['no_tests', 'scheduled', 'not_run'].includes(circleStatus)) {
+        color = 'yellow'
+        shieldsStatus = circleStatus.replace('_', ' ')
+        return { status: shieldsStatus, color: color }
+      } else {
+        color = 'lightgrey'
+        shieldsStatus = circleStatus.replace('_', ' ')
+        return { status: shieldsStatus, color: color }
+      }
+    }
+
+    if (passCount === data.length) {
+      color = 'brightgreen'
+      shieldsStatus = 'passing'
+    }
+
+    return { status: shieldsStatus, color: color }
+  }
+
+  async handle({ token, type, userRepo, branch }) {
+    const json = await this.fetch({ token, type, userRepo, branch })
+    const { status, color } = this.transform(json)
+    return this.constructor.render({ status, color })
```

Comment:
As I write and review more code using this `fetch()`/`render()` pattern, I think we're really close to a point where we should update the docs to formalise this pattern. I think if #1922 (another complex service 'family') goes in without any major change to the base classes (beyond adding the `Deprecated` exception), we should make docs the next priority.